cpp/src/Slice/SliceUtil.cpp

```diff
@@ -470,59 +470,49 @@ Slice::checkIdentifier(const string& id)
         name = id;
     }
 
+    // Validity is only determined by invalid syntax that triggers errors, not warnings.
     bool isValid = true;
+
     // check the identifier for reserved suffixes
     static const string suffixBlacklist[] = { "Helper", "Holder", "Prx", "Ptr" };
     for(size_t i = 0; i < sizeof(suffixBlacklist) / sizeof(*suffixBlacklist); ++i)
     {
         if(name.find(suffixBlacklist[i], name.size() - suffixBlacklist[i].size()) != string::npos)
         {
-            unit->error("illegal identifier `" + name + "': `" + suffixBlacklist[i] + "' suffix is reserved");
-            isValid = false;
+            unit->warning(ReservedIdentifiers, "`" + suffixBlacklist[i] +
+                                               "' suffix is reserved by Ice: the identifier `" + name +
+                                               "' may conflict with generated code");
         }
     }
 
+    // Check the identifier for illegal ice prefixes
+    if(unit->currentIncludeLevel() == 0 && ciequals(name.substr(0, 3), "ice"))
+    {
+        unit->warning(ReservedIdentifiers, "`" + name.substr(0, 3) + "' prefix is reserved by Ice: the identifier `" +
+                                           name + "' may conflict with generated code");
+    }
+
     // check the identifier for illegal underscores
     size_t index = name.find('_');
     if(index == 0)
     {
-        unit->error("illegal leading underscore in identifier `" + name + "'");
+        unit->error("illegal identifier: `" + name + "', leading underscores are reserved by Ice");
```

Comment:
I prefer the original error message (illegal leading underscore in identifier). Let's reserve "reserved" for the identifiers we only warn about.

Reply:
I find the new message explains why the error is happening better than the original does in my opinion, and it certainly sounds more natural to read as well. Why not call it 'reserved' when that's what it is?

Reply:
These identifiers are not reserved, they are illegal. Identifiers that are merely reserved trigger a warning that can be suppressed.